File: src/layup/propagate.py
"""Move fitted orbits to a different reference epoch (issue #578).

``convert`` changes the *parameterization* of an orbit at a fixed epoch;
this changes the epoch. The two are complementary and neither substitutes for
the other.

The state is integrated with ASSIST, the same dynamics the fitter uses, and the
covariance is carried with it as ``C' = Phi C Phi^T`` where ``Phi`` is the
state-transition matrix over the interval.

No light-time correction is applied. Propagation to an epoch is a dynamical
operation; :mod:`layup.predict` applies light time because it computes an
observable, and conflating the two would be a subtle and expensive error.

Why this exists: Layup already emits two epoch conventions and could not
reconcile them. A cold-started fit is referenced to the middle observation of
its IOD triplet, so it sits inside its own arc; a warm-started fit inherits the
epoch of its initial guess, so it sits whereever the catalogue that seeded it
does. Comparing the two, or comparing either against a reference solution
published at a third epoch, requires moving one of them.
"""

import logging

import numpy as np

from layup.routines import get_ephem, propagate_state
from layup.utilities.data_processing_utilities import parse_fit_result
from layup.utilities.cache_location import default_cache_dir

logger = logging.getLogger(__name__)

__all__ = ["propagate", "propagate_row"]

MJD_TO_JD = 2400000.5


def _cache(cache_dir):
    return str(default_cache_dir()) if cache_dir is None else str(cache_dir)


def propagate_row(fit_row, target_epoch_mjd_tdb, cache_dir=None, return_stm=False):
    """Propagate one fit result to ``target_epoch_mjd_tdb``.

    Parameters
    ----------
    fit_row : numpy structured array
        One row of ``orbitfit`` output. Its covariance columns are carried if
        present; a row without them propagates the state alone.
    target_epoch_mjd_tdb : float
        Target epoch, MJD TDB -- the same units as the ``epochMJD_TDB`` column.
    cache_dir : str, optional
        Kernel cache. Defaults to Layup's.
    return_stm : bool
        Also return the 6x6 state-transition matrix.

    Returns
    -------
    numpy structured array
        A copy of ``fit_row`` at the new epoch. ⚠️ **Check the flag.** If the
        propagation fails -- most often a target outside the ephemeris
        coverage, 1550 to 2650 -- the row comes back with ``flag != 0`` at its
        ORIGINAL epoch and state, so a caller that ignores the flag gets an
        unchanged orbit rather than a plausible wrong one.
    """
    row = np.atleast_1d(fit_row).copy()
    if len(row) != 1:
        raise ValueError(f"propagate_row takes a single row, got {len(row)}")

    fit = parse_fit_result(row)
    out, stm = propagate_state(get_ephem(_cache(cache_dir)), fit, float(target_epoch_mjd_tdb) + MJD_TO_JD)

    if out.flag == 0:
        for i, c in enumerate(("x", "y", "z", "xdot", "ydot", "zdot")):
            if c in row.dtype.names:
                row[c] = out.state[i]
        if "epochMJD_TDB" in row.dtype.names:
            row["epochMJD_TDB"] = out.epoch - MJD_TO_JD
        cov = list(out.cov)
        for i in range(6):
            for j in range(6):
                name = f"cov_{i}_{j}"
                if name in row.dtype.names:
                    row[name] = cov[i * 6 + j]
    else:
        if "flag" in row.dtype.names:
            row["flag"] = out.flag
        logger.warning(
            "Propagation to MJD %.6f failed; returning the orbit unchanged at MJD %.6f. "
            "The most common cause is a target outside the ephemeris coverage (1550-2650).",
            float(target_epoch_mjd_tdb),
            float(row["epochMJD_TDB"][0]) if "epochMJD_TDB" in row.dtype.names else float("nan"),
        )

    if return_stm:
        return row, np.asarray(list(stm), dtype=float).reshape(6, 6)
    return row


def propagate(fit_results, target_epoch_mjd_tdb, cache_dir=None):
    """Propagate every row of ``fit_results`` to a common epoch.

    The point of a *common* epoch: two orbits cannot be compared, differenced or
    averaged unless they are referenced to the same time. Differencing states at
    epochs that differ by even a minute is a mistake that leaves no trace in the
    output -- 69 seconds is about 1,000 km for a main-belt object and 2,000 km
    for a fast near-Earth one.

    Rows that fail to propagate are returned unchanged with ``flag != 0``; they
    are counted in a warning rather than dropped, so the output has the same
    length and order as the input.
    """
    data = np.atleast_1d(fit_results)
    out = data.copy()
    failed = 0
    for i in range(len(data)):
        row = propagate_row(data[i : i + 1], target_epoch_mjd_tdb, cache_dir=cache_dir)
        out[i] = row[0]
        if "flag" in out.dtype.names and out[i]["flag"] != 0:
            failed += 1
    if failed:
        logger.warning(
            "%d of %d orbits could not be propagated to MJD %.6f and are unchanged.",
            failed,
            len(data),
            float(target_epoch_mjd_tdb),
        )
    return out

File: src/lib/orbit_fit/propagate.cpp
// Propagate a fitted state, and its covariance, to a new reference epoch
// (issue #578).
//
// `convert` changes the PARAMETERIZATION of an orbit at a fixed epoch; nothing
// changed the epoch itself. The integration machinery was already here --
// predict.cpp drives assist_integrate_or_interpolate -- but every entry point
// returned sky quantities, so a state at a different epoch had to be obtained
// by writing the integration again outside the package.
//
// This is a purely dynamical operation. It deliberately applies NO light-time
// correction: `predict` does, because it computes an observable, and
// conflating the two would be a subtle and expensive error.
//
// The covariance is carried by the state-transition matrix, C' = Phi C Phi^T,
// with Phi obtained from six variational particles seeded along the state
// axes -- the same mechanism the fitter uses for its partials.
//
// Included from orbit_fit.cpp at the bottom of `namespace orbit_fit`, so that
// FitResult, add_variational_particles and apply_ias15_adaptive_mode are all
// in scope. It is not a standalone translation unit.

    // Returns the propagated fit and the 6x6 state-transition matrix (row-major,
    // Phi[i][j] = d state_i(t) / d state_j(t0)).
    //
    // On failure the returned FitResult carries flag != 0 and the input epoch,
    // so a caller that ignores the flag gets an unchanged orbit rather than a
    // plausible wrong one.
    std::pair<FitResult, std::array<double, 36>>
    propagate_state(struct assist_ephem *ephem, FitResult fit, double target_epoch)
    {
        FitResult out = fit;
        std::array<double, 36> stm{};

        // Nothing to do, and doing it anyway would burn an integration.
        if (target_epoch == fit.epoch)
        {
            for (int i = 0; i < 6; i++) stm[i * 6 + i] = 1.0;
            return {out, stm};
        }

        struct reb_simulation *r = reb_simulation_create();
        r->t = fit.epoch - ephem->jd_ref;

        struct reb_particle p0 = {0};
        p0.x = fit.state[0]; p0.y = fit.state[1]; p0.z = fit.state[2];
        p0.vx = fit.state[3]; p0.vy = fit.state[4]; p0.vz = fit.state[5];
        reb_simulation_add(r, p0);

        // Six variational particles seeded with the unit state axes: their
        // states at the target epoch ARE the columns of Phi. `np = 0` is passed
        // explicitly -- REBOUND's default of -1 would leave ASSIST filling no
        // variational acceleration at all, and Phi would come back as the
        // identity with no error raised.
        int var = 0;
        add_variational_particles(r, 0, &var, 0);

        struct assist_extras *ax = assist_attach(r, ephem);
        apply_ias15_adaptive_mode(r);  // after attach: ASSIST forces mode 1

        assist_integrate_or_interpolate(ax, target_epoch - ephem->jd_ref);

        if (r->status == REB_STATUS_GENERIC_ERROR)
        {
            // Outside ephemeris coverage, or the integration failed. Report the
            // input unchanged with a non-zero flag rather than a wrong state.
            out.flag = 1;
            assist_free(ax);
            reb_simulation_free(r);
            return {out, stm};
        }

        const struct reb_particle pf = r->particles[0];
        out.state[0] = pf.x; out.state[1] = pf.y; out.state[2] = pf.z;
        out.state[3] = pf.vx; out.state[4] = pf.vy; out.state[5] = pf.vz;
        out.epoch = target_epoch;

        for (int j = 0; j < 6; j++)
        {
            const struct reb_particle v = r->particles[var + j];
            stm[0 * 6 + j] = v.x;  stm[1 * 6 + j] = v.y;  stm[2 * 6 + j] = v.z;
            stm[3 * 6 + j] = v.vx; stm[4 * 6 + j] = v.vy; stm[5 * 6 + j] = v.vz;
        }

        // C' = Phi C Phi^T. A zeroed input covariance stays zeroed, which is
        // what an unconverged fit carries (#547) and is the right answer for it.
        Eigen::Matrix<double, 6, 6> Phi, C;
        for (int i = 0; i < 6; i++)
            for (int j = 0; j < 6; j++)
            {
                Phi(i, j) = stm[i * 6 + j];
                C(i, j) = fit.cov[i * 6 + j];
            }
        const Eigen::Matrix<double, 6, 6> Cn = Phi * C * Phi.transpose();
        for (int i = 0; i < 6; i++)
            for (int j = 0; j < 6; j++)
                out.cov[i * 6 + j] = Cn(i, j);

        assist_free(ax);
        reb_simulation_free(r);
        return {out, stm};
    }

    static void propagate_bindings(py::module &m)
    {
        m.def("propagate_state", &orbit_fit::propagate_state,
              R"pbdoc(
                Propagate a fitted state and its covariance to a new reference
                epoch (issue #578).

                Returns ``(fit, stm)``: the FitResult re-referenced to
                ``target_epoch``, and the 6x6 state-transition matrix as 36
                row-major values, ``stm[i*6+j] = d state_i(t) / d state_j(t0)``.

                The covariance is carried as ``C' = Phi C Phi^T``. No light-time
                correction is applied -- this is a dynamical operation, not an
                observable. On failure, including a target outside the
                ephemeris coverage, the returned fit carries ``flag != 0`` and
                the ORIGINAL epoch and state.
              )pbdoc",
              py::arg("ephem"), py::arg("fit"), py::arg("target_epoch"));
    }
